### PR TITLE
Add SAN to certificate and include related tests

### PR DESCRIPTION
**Pull request by giovanni-riela**
#### Motivation and Context

This PR introduces the Subject Alternative Name (SAN) extension to the Wallet Provider's X.509 certificates to ensure compliance with modern security standards and IT-Wallet ecosystem requirements.
Previously, the leaf certificate only utilized the Common Name (CN); this change adds a DNS-type SAN entry matching the provider's host. To support this, a new utility was added to detect the presence of the SAN extension, and the certificate loading logic was updated to automatically regenerate any existing cached certificates that lack this required extension. Unit tests have been included to verify both the correct generation of the extension and the validation logic.

JIRA Ticket: WLEO-1247

### CODE REVIEW

File: src/logic/pem.ts [hunk 1]
```diff
@@ -212,6 +214,15 @@ export async function createSignedCertificate(
   return cert;
 }
 
+/**
+ * Returns `true` when the certificate (base64-DER) carries a
+ * SubjectAlternativeName extension (OID 2.5.29.17).
+ */
+export function hasSanExtension(certDerBase64: string): boolean {
```
Comment (manuraf) [inline src/logic/pem.ts:223]: `hasSanExtension` currently calls `new x509.X509Certificate(certDerBase64)` without any error guard. If the input is empty, corrupted, or accidentally a full PEM string (with `-----BEGIN CERTIFICATE-----` headers) instead of stripped base64-DER, the constructor throws and the exception propagates unchecked to the caller. Since the function signature promises `boolean` semantics, callers — including the cache-validity guard in `loadWalletProviderCertificate` — won't naturally wrap it in a `try/catch`.

You could wrap the body with a try/catch and return `false` on failure. This keeps the boolean contract intact and, as a bonus, treats a corrupted on-disk certificate as "SAN missing", which automatically triggers regeneration:

```typescript
export function hasSanExtension(certDerBase64: string): boolean {
  if (!certDerBase64) return false;
  try {
    const cert = new x509.X509Certificate(certDerBase64);
    return !!cert.getExtension("2.5.29.17");
  } catch {
    return false; // treat unparseable cert as missing SAN → triggers regen
  }
}
```

It would also be worth applying the same guard to `hasX509CertificateExpired` for consistency, since both functions share the same latent throw pattern.

Comment (manuraf) [inline src/logic/pem.ts:223]: `hasSanExtension` currently calls `new x509.X509Certificate(certDerBase64)` without any error guard. If the input is empty, corrupted, or accidentally a full PEM string (with `-----BEGIN CERTIFICATE-----` headers) instead of stripped base64-DER, the constructor throws and the exception propagates unchecked to the caller. Since the function signature promises `boolean` semantics, callers — including the cache-validity guard in `loadWalletProviderCertificate` — won't naturally wrap it in a `try/catch`.

You could wrap the body with a try/catch and return `false` on failure. This keeps the boolean contract intact and, as a bonus, treats a corrupted on-disk certificate as "SAN missing", which automatically triggers regeneration:

```typescript
export function hasSanExtension(certDerBase64: string): boolean {
  if (!certDerBase64) return false;
  try {
    const cert = new x509.X509Certificate(certDerBase64);
    return !!cert.getExtension("2.5.29.17");
  } catch {
    return false; // treat unparseable cert as missing SAN → triggers regen
  }
}
```

It would also be worth applying the same guard to `hasX509CertificateExpired` for consistency, since both functions share the same latent throw pattern.

File: src/logic/pem.ts [hunk 2]
```diff
@@ -212,6 +214,15 @@ export async function createSignedCertificate(
   return cert;
 }
 
+/**
+ * Returns `true` when the certificate (base64-DER) carries a
```
Comment (manuraf) [inline src/logic/pem.ts:220]: The OID `"2.5.29.17"` is currently duplicated as a raw string literal in the implementation and again in the test file via a local `const SAN_OID`. A typo here would be a silent runtime bug — `getExtension` would always return `undefined` with no compile-time signal.

You could export a named constant directly from `pem.ts` and use it in the function body:

```typescript
// pem.ts
export const OID_SUBJECT_ALTERNATIVE_NAME = "2.5.29.17" as const;

export function hasSanExtension(certDerBase64: string): boolean {
  // ...
  return !!cert.getExtension(OID_SUBJECT_ALTERNATIVE_NAME);
}
```

Then in the test file you can drop the local `SAN_OID` and import the constant instead:

```typescript
// wallet-provider-cert.unit.spec.ts — remove SAN_OID, import from pem
import { createSignedCertificate, hasSanExtension, OID_SUBJECT_ALTERNATIVE_NAME } from "@/logic/pem";
// Remove: const SAN_OID = "2.5.29.17";
// Update: cert.getExtension(OID_SUBJECT_ALTERNATIVE_NAME)
```

This gives a single source of truth and makes any future consumer of `pem.ts` automatically discoverable.

File: src/logic/wallet-provider.ts [hunk 3]
```diff
@@ -1,3 +1,4 @@
+import * as x509 from "@peculiar/x509";
```
Comment (manuraf) [inline src/logic/wallet-provider.ts:1]: `intermediateKeyPair` contains the `"d"` (private key) component of an intermediate CA with `keyCertSign` usage — it can sign leaf certificates trusted within this wallet's PKI. Writing it with `writeFileSync` and no `mode` option leaves the file world-readable (`0644` under the default `umask 022`).

You could add `{ mode: 0o600 }` here, consistent with the pattern already used in `pem.ts:87` for other key files:

```typescript
writeFileSync(
  intermediateJwksPath,
  JSON.stringify(intermediateKeyPair),
  { mode: 0o600 },
);
```

This is a one-line fix with no risk. Worth noting there is a separate open question (todo #007) about whether this write is ever read back — if it turns out to be dead code, removing it entirely would be the cleaner resolution.

File: tests/unit/wallet-provider-cert.unit.spec.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+import { describe, expect, it } from "vitest";
+
+import { createKeys } from "@/logic/jwk";
+import { createSignedCertificate, hasSanExtension } from "@/logic/pem";
+import { LOCAL_WP_HOST } from "@/servers/wp-server";
+
+const SAN_OID = "2.5.29.17";
+
+describe("wallet_provider_cert SAN", () => {
+  it("leaf cert created with SAN contains the expected dNSName entry", async () => {
```
Comment (manuraf) [inline tests/unit/wallet-provider-cert.unit.spec.ts]: Tests 1 and 2 share identical setup code — each independently calls `createKeys()` twice and `createSignedCertificate()` once, running 2 extra async operations that serve no purpose since the resulting certificates are equivalent.

You could consolidate the shared setup into a `beforeAll` block, creating the certificates once and referencing them across all three tests. This keeps each `it` as a pure assertion with no setup noise:

```typescript
describe("wallet_provider_cert SAN", () => {
  let certWithSan: x509.X509Certificate;
  let certWithSanDerBase64: string;
  let certWithoutSanDerBase64: string;

  beforeAll(async () => {
    const issuerKp = await createKeys();
    const subjectKp = await createKeys();
    certWithSan = await createSignedCertificate(issuerKp, "CN=TestIssuer", subjectKp, `CN=${LOCAL_WP_HOST}`, false, [
      new x509.SubjectAlternativeNameExtension([{ type: "dns", value: LOCAL_WP_HOST }], false),
    ]);
    certWithSanDerBase64 = Buffer.from(certWithSan.rawData).toString("base64");

    const issuerKp2 = await createKeys();
    const subjectKp2 = await createKeys();
    const certWithoutSan = await createSignedCertificate(issuerKp2, "CN=TestIssuer", subjectKp2, `CN=${LOCAL_WP_HOST}`, false);
    certWithoutSanDerBase64 = Buffer.from(certWithoutSan.rawData).toString("base64");
  });

  it("leaf cert created with SAN extension carries OID 2.5.29.17", () => {
    expect(certWithSan.getExtension(OID_SUBJECT_ALTERNATIVE_NAME), "SAN extension must be present").not.toBeNull();
  });

  it("hasSanExtension returns true for cert with SAN", () => {
    expect(hasSanExtension(certWithSanDerBase64), "hasSanExtension should return true when SAN is present").toBe(true);
  });

  it("hasSanExtension returns false for cert without SAN", () => {
    expect(hasSanExtension(certWithoutSanDerBase64), "hasSanExtension should return false when SAN is absent").toBe(false);
  });
});
```

As a side note, Test 1 asserts that `cert.getExtension(...)` is non-null after you passed the extension in — that's testing `@peculiar/x509` internals rather than your own code. If the team doesn't need explicit regression coverage of library behavior, it could be dropped. Either way, the `beforeAll` refactor is independent of that decision.

File: tests/unit/wallet-provider-cert.unit.spec.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      `CN=${LOCAL_WP_HOST}`,
+      false,
+    );
+
+    const certDerBase64 = Buffer.from(cert.rawData).toString("base64");
+    expect(
+      hasSanExtension(certDerBase64),
+      "hasSanExtension should return false when SAN is absent",
+    ).toBe(false);
+  });
```
Comment (manuraf) [inline tests/unit/wallet-provider-cert.unit.spec.ts:82]: Once the defensive `try/catch` is in place inside `hasSanExtension` (see the related fix), it would be worth adding a couple of edge-case tests here as regression guards. Without them, a future refactor could silently drop the defensive behaviour and the breakage would go completely undetected.

You could append these two `it` blocks to the existing spec:

```typescript
it("hasSanExtension returns false for an empty string", () => {
  expect(
    hasSanExtension(""),
    "empty string should not throw and should return false",
  ).toBe(false);
});

it("hasSanExtension returns false for malformed base64 input", () => {
  expect(
    hasSanExtension("not-valid-der-base64!!!"),
    "malformed input should not throw and should return false",
  ).toBe(false);
});
```

They cover the two most likely bad inputs (empty string and non-base64 DER) and are low-effort to write — roughly 10 minutes of work.

File: src/logic/wallet-provider.ts [hunk 6]
```diff
@@ -106,6 +102,12 @@ export async function loadWalletProviderCertificate(
     providerKeyPair,
     `CN=${providerDomain}`,
     false,
```
Comment (flaviodelgrosso) [inline src/logic/wallet-provider.ts:104]: This is adding a DNS SAN for the host only, but the wallet-provider certificate chain needs the SAN to be a URI that matches the federation entity identifier used in the provider entity configuration (`iss`/`sub`). As written, the generated x5c can carry the extension and still fail that binding. I think this should emit a SAN URI for the full wallet provider base URL and cover that exact value in the tests. What do you think?

Comment (giovanni-riela) [reply src/logic/wallet-provider.ts:104]: I pushed the change, let me know if that was what you had in mind

**Review by flaviodelgrosso — APPROVED** [review]

**Review by manuraf — APPROVED** [review]
